### PR TITLE
Un-publish on Workspace Delete

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses 
https://broadinstitute.atlassian.net/browse/GAWB-2352
Requires corresponding UI PR: https://github.com/broadinstitute/firecloud-ui/pull/931

## Changes:
Update the delete workspace logic to include un-publishing from Library if necessary.

## Testing:
* Delete a normal, un-published workspace and ensure no regressions
* Delete a published workspace and ensure that it does not show up in Library.
* Try to delete a published workspace with users that are not allowed.
  * Workspace "READER"
  * Workspace "OWNER" but not "CATALOG" permissions
  * "CATALOG" user, but not "OWNER"

---

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](../CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](../CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [x] Get two thumbsworth of review and PO signoff if necessary
- [x] Verify all tests go green
- [x] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [x] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: this duplicates a bunch of logic with `LibraryService.setWorkspaceIsPublished`. Could you call that method instead of duplicating here? You might have to move `setWorkspaceIsPublished` into `WorkspacePublishingSupport` or `LibraryServiceSupport` to achieve the reuse.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: by replaying the error message from rawls, the end user - in the case where the user is owner but not curator - gets a message "You must be a curator and either be an owner or have catalog with read+." That's correct, but not necessarily helpful. Maybe change to something like `s"You cannot delete this workspace: ${e.getMessage}"`

**Review by davidangb — COMMENTED** [review]
two comments inline.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: I started out on that road, but that method is doing much more than what I want to do and is completing the request in cases that I don't want it to. I think I can extract out the last `else` condition in that method to look like what I have.

**Review by MattPutnam — APPROVED** [review]
👍 besides what David said

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: @davidangb - Since `setWorkspaceIsPublished` has a good chunk of preliminary logic checking that this feature doesn't need to go through, I ended up extracting what I needed from `setWorkspaceIsPublished` into another method in `WorkspacePublishingSupport`. Please have a second look.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: The other way I've seen this kind of logic flow handled is:
```
      val unpublishFuture = if (isPublished(ws))
        setWorkspacePublishedStatus(ws.workspace, publishArg = false, rawlsDAO, ontologyDAO, searchDAO)
      else
        Future.successful(ws)

      unpublishFuture flatMap { _ =>
        rawlsDAO.deleteWorkspace(ns, name) map { wsResponse =>
          RequestComplete(wsResponse.copy(message = Some(wsResponse.message.getOrElse("") + unPublishSuccessMessage(ns, name))))
        } recover {
          case e: FireCloudExceptionWithErrorReport => RequestComplete(e.errorReport.statusCode.getOrElse(InternalServerError), ErrorReport(message = s"You cannot delete this workspace: ${e.errorReport.message}"))
          case e: Throwable => RequestComplete(InternalServerError, ErrorReport(message = s"You cannot delete this workspace: ${e.getMessage}"))
        }
      }
```
I think it's slightly cleaner, but I'm not officially requesting a change for this PR; your call if you want to adopt different syntax or not.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala]: scope creep warning!

If the `updateLibraryAttributes ` succeeds, but the `Try(publishDocument)` or `Try(removeDocument)` fails, we're out of sync. Now that you have these in a `Try`, it would be easy enough to look for failures and reset the published attribute. This is GAWB-1683, so feel free to push it off.

**Review by davidangb — APPROVED** [review]
two comments inline; both optional. I like the refactoring! :+1:

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: 👍  I like that approach better, more granular and less nest-y.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspacePublishingSupport.scala]: Yes, that is a problem. I'm going to add some error logging around this so we can get alerted to it, but leave the real fix for 1683.

**Review by davidangb — APPROVED** [review]
:+1:

**Comment by rushtong** [comment]
Getting totally unrelated travis errors. Need to sort it out before merging.
```
[info] HttpOntologyDAO concurrency
[info] - should obey the spray max-clients throttle *** FAILED ***
[info]   Expected 8, but got 4 request count should have reached its maximum throttle limit (HttpOntologyDAOConcurrencySpec.scala:68)
```

**Pull request merged**
